File: app/src/main/java/pl/gocards/ui/cards/xml/list/select/SelectCardViewHolder.java
package pl.gocards.ui.cards.xml.list.select;

import android.view.MotionEvent;

import androidx.annotation.NonNull;

import com.google.android.material.color.MaterialColors;

import pl.gocards.R;
import pl.gocards.databinding.ItemCardBinding;
import pl.gocards.ui.cards.xml.list.drag_swipe.DragSwipeCardViewHolder;
import pl.gocards.ui.cards.xml.list.select.popup.menu.NotSelectedCardPopupMenu;
import pl.gocards.ui.cards.xml.list.select.popup.menu.SelectCardPopupMenu;
import pl.gocards.ui.cards.xml.list.select.popup.menu.SelectedCardPopupMenu;

/**
 * C_R_08 Select (cut) the card
 *
 * @author Grzegorz Ziemski
 */
public class SelectCardViewHolder extends DragSwipeCardViewHolder {

    public SelectCardViewHolder(@NonNull ItemCardBinding binding, @NonNull SelectListCardsAdapter adapter) {
        super(binding, adapter);
    }

    /**
     * C_02_01 When no card is selected and tap on the card, show the popup menu.
     */
    @Override
    protected void showSingleTapMenu() {
        new SelectCardPopupMenu(requireActivity(), this).showPopupMenu();
    }

    /**
     * C_02_04 When any card is selected and long pressing on the card, show the selected popup menu.
     */
    public void showSelectPopupMenu() {
        if (getAdapter().isCardSelected(getBindingAdapterPosition())) {
            new SelectedCardPopupMenu(requireActivity(), this).showPopupMenu();
        } else {
            new NotSelectedCardPopupMenu(requireActivity(), this).showPopupMenu();
        }
    }

    /* -----------------------------------------------------------------------------------------
     * Implementation of GestureDetector
     * ----------------------------------------------------------------------------------------- */

    /**
     * C_02_03 When any card is selected and tap on the card, select or unselect the card.
     */
    @Override
    public boolean onSingleTapUp(@NonNull MotionEvent e) {
        if (getAdapter().isSelectionMode()) {
            getAdapter().onCardInvertSelect(this);
            return false;
        } else {
            return super.onSingleTapUp(e);
        }
    }

    /* -----------------------------------------------------------------------------------------
     * Change the look of the view.
     * ----------------------------------------------------------------------------------------- */

    @Override
    public void unfocusItemView() {
        // Check that the card has not been previously selected.
        int position = getBindingAdapterPosition();
        if (-1 < position && position < getAdapter().getItemCount()) {
            if (getAdapter().isCardSelected(position)) {
                selectItemView();
            } else {
                super.unfocusItemView();
            }
        }
    }

    protected void selectItemView() {
        this.itemView.setSelected(true);
        this.itemView.setBackgroundColor(
                MaterialColors.getColor(this.itemView, R.attr.colorItemSelected)
        );
    }

    /* -----------------------------------------------------------------------------------------
     * Gets/Sets
     * ----------------------------------------------------------------------------------------- */

    @NonNull
    protected SelectListCardsAdapter getAdapter() {
        return (SelectListCardsAdapter) super.getAdapter();
    }

    @NonNull
    @Override
    protected SelectListCardsActivity requireActivity() {
        return (SelectListCardsActivity) super.requireActivity();
    }
}

File: app/src/main/java/pl/gocards/ui/cards/kt/list/select/SelectCardViewHolder.kt
package pl.gocards.ui.cards.kt.list.select

import android.view.MotionEvent
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.toArgb
import pl.gocards.databinding.ItemCardBinding
import pl.gocards.ui.cards.kt.list.ListCardsActivity
import pl.gocards.ui.cards.kt.list.display.CardViewHolder
import pl.gocards.ui.cards.kt.list.display.ListCardsPopupMenu
import pl.gocards.ui.kt.theme.ExtendedColors

/**
 * @author Grzegorz Ziemski
 */
class SelectCardViewHolder(
    binding: ItemCardBinding,
    colors: ExtendedColors,
    activity: ListCardsActivity,
    val adapter: SelectListCardsAdapter
): CardViewHolder(binding, colors, activity) {

    /* -----------------------------------------------------------------------------------------
     * Change the look of the view.
     * ----------------------------------------------------------------------------------------- */

    fun select() {
        // itemView.isSelected = true
        itemView.setBackgroundColor(colors.colorItemSelected.toArgb())
    }

    /* -----------------------------------------------------------------------------------------
     * GestureDetector
     * ----------------------------------------------------------------------------------------- */

    /**
     * C_02_03 When any card is selected and tap on the card, select or unselect the card.
     */
    override fun onSingleTapUp(e: MotionEvent): Boolean {
        return if (adapter.isSelectionMode()) {
            adapter.invertSelection(this)
            false
        } else {
            super.onSingleTapUp(e)
        }
    }

    @Composable
    override fun CreatePopupMenu(onDismiss: () -> Unit) {
        if (adapter.isSelectionMode()) {
            SelectListCardsPopupMenu(adapter, this, onDismiss)
        } else {
            ListCardsPopupMenu(adapter, this, onDismiss)
        }
    }
}